### PR TITLE
 add docker-bench-security results parser, add parser tests

### PR DESCRIPTION
**Pull request by drJabber**
Added OWASP [docker-bench-security](https://github.com/docker/docker-bench-security) results parser, add parser tests
docker-bench-security is a script that make tests based on [CIS Docker Benchmark](https://www.cisecurity.org/benchmark/docker/)

### CODE REVIEW

File: unittests/tools/test_dockerbench_parser.py [hunk 1]
```diff
@@ -0,0 +1,30 @@
...
+        findings = parser.get_findings(testfile, Test())
+        self.assertEqual(1, len(findings))
+
+    def test_parse_file_with_multiple_vuln_has_multiple_findings(self):
+        testfile = open(
+            get_unit_tests_path() + "/scans/dockerbench/docker-bench-report-many-vulns.json"
+        )
+        parser = DockerBenchParser()
+        findings = parser.get_findings(testfile, Test())
+        self.assertTrue(len(findings) == 50)
```
Comment (damiencarol) [inline unittests/tools/test_dockerbench_parser.py:30]: Please add more check on the findings

**Review by damiencarol — CHANGES_REQUESTED** [review]

File: unittests/tools/test_dockerbench_parser.py (hunk 1 above)
Comment (drJabber) [reply unittests/tools/test_dockerbench_parser.py:30]: @damiencarol , thank you,  added some extra assertions in this test

Comment (drJabber) [reply unittests/tools/test_dockerbench_parser.py:30]: @damiencarol , fixed failed unit tests.

File: dojo/tools/dockerbench/parser.py [hunk 2]
```diff
@@ -0,0 +1,115 @@
...
+        description += '\n'
+        description += 'Audit: {}\n'.format(vuln['audit'])
+    if 'expected_result' in vuln and vuln['expected_result'] != '':
+        description += '\n'
+        description += 'Expected result: {}\n'.format(vuln['expected_result'])
+    if 'actual_value' in vuln and vuln['actual_value'] != '':
+        description += '\n'
+        description += 'Actual value: {}\n'.format(vuln['actual_value'])
+
+    mitigation = vuln.get('remediation', '')
```
Comment (damiencarol) [inline dojo/tools/dockerbench/parser.py:99]:
```suggestion
    mitigation = vuln.get('remediation')
```
Mitigation attribute is not mandatory

File: dojo/tools/dockerbench/parser.py [hunk 3]
```diff
@@ -0,0 +1,115 @@
...
+    vuln_id_from_tool = test_number
+
+    finding = Finding(title=title,
+                      test=test,
+                      description=description,
+                      severity=severity,
+                      mitigation=mitigation,
+                      vuln_id_from_tool=vuln_id_from_tool,
+                      static_finding=True,
+                      dynamic_finding=False)
```
Comment (damiencarol) [inline dojo/tools/dockerbench/parser.py:113]: The date attribute is missing, you have the date in the `end` attribute of the report.

File: dojo/tools/dockerbench/parser.py [hunk 4]
```diff
@@ -0,0 +1,115 @@
...
+        description += 'desc: {}\n'.format(vuln['desc'])
+    if 'details' in vuln:
+        description += vuln['details']
+    if 'audit' in vuln:
+        description += '\n'
+        description += 'Audit: {}\n'.format(vuln['audit'])
+    if 'expected_result' in vuln and vuln['expected_result'] != '':
+        description += '\n'
+        description += 'Expected result: {}\n'.format(vuln['expected_result'])
+    if 'actual_value' in vuln and vuln['actual_value'] != '':
```
Comment (damiencarol) [inline dojo/tools/dockerbench/parser.py:95]:
```suggestion
    if vuln.get('actual_value'):
```

File: dojo/tools/dockerbench/parser.py [hunk 5]
```diff
@@ -0,0 +1,115 @@
...
+    if status.upper() == 'FAIL':
+        severity = 'Critical'
+    elif status.upper() == 'WARN' and '(Manual)' not in reason:
+        severity = 'High'
+    elif status.upper() == 'INFO' and '(Manual)' not in reason:
+        severity = 'Low'
+    elif status.upper() == 'NOTE' and '(Manual)' not in reason:
+        severity = 'Info'
+    else:
+        return None
```
Comment (damiencarol) [inline dojo/tools/dockerbench/parser.py:75]: this is not good, severity is mandatory.

File: dojo/tools/dockerbench/parser.py [hunk 6]
```diff
@@ -0,0 +1,115 @@
...
+
+    test_number = vuln.get('id', 'Test number not found')
```
Comment (damiencarol) [inline dojo/tools/dockerbench/parser.py:77]: I think we should consider that `id` is require for this parser and organize the code like this.

**Review by damiencarol — CHANGES_REQUESTED** [review]
It's a good first shot.
Some stuffs are missing (like severity)
Also you should add de-duplication default configuration in dojo/settings/dist.settings.py

File: dojo/tools/dockerbench/parser.py (hunk 2 above)
Comment (drJabber) [reply dojo/tools/dockerbench/parser.py:99]: Yes, really, thank you

File: dojo/tools/dockerbench/parser.py (hunk 4 above)
Comment (drJabber) [reply dojo/tools/dockerbench/parser.py:95]: Ok, the same as remediation

File: dojo/tools/dockerbench/parser.py (hunk 5 above)
Comment (drJabber) [reply dojo/tools/dockerbench/parser.py:75]: Thank you, but severity is mandatory when we add the item to result findings. 
Here I dont assign None to severity attribute of finding, here I return None from get_item function (e.g when status is PASS)- I dont create finding in this situation and dont return it to Dojo in get_findings result. 
So I think I can return None when the docker-security-bench result item is not a finding.

File: dojo/tools/dockerbench/parser.py (hunk 3 above)
Comment (drJabber) [reply dojo/tools/dockerbench/parser.py:113]: thank you, test date is important. but it does not contain TZ. 
so, I decode it as a local timestamp - datetime.fromtimestamp

File: dojo/tools/dockerbench/parser.py (hunk 6 above)
Comment (drJabber) [reply dojo/tools/dockerbench/parser.py:77]: I agree, but I think code should not crash when there no id (user could edit file manually),
i tried to fix and now:
vuln_id_from_tool will be id or None, 
and title will be "<id> - <desc>", or "No test number - <desc>"

**Comment by drJabber** [comment]
> It's a good first shot. Some stuffs are missing (like severity) Also you should add de-duplication default configuration in dojo/settings/dist.settings.py

Thank you, @damiencarol , tried to fix all, also added dedupe config to dist.settings.py

File: dojo/tools/dockerbench/parser.py (hunk 6 above)
Comment (damiencarol) [reply dojo/tools/dockerbench/parser.py]: this is good, will check again the code

File: dojo/settings/settings.dist.py [hunk 7]
```diff
@@ -1123,6 +1123,7 @@ def saml2_attrib_map_format(dict):
     'Hydra Scan': ['title', 'description'],
     'DrHeader JSON Importer': ['title', 'description'],
     'PWN SAST': ['title', 'description'],
+    'docker-bench-security Scan': ['title', 'description'],
```
Comment (damiencarol) [inline dojo/settings/settings.dist.py:1126]: I think we should use vuln_id_from_tool here

Comment (drJabber) [reply dojo/settings/settings.dist.py:1126]: ok, thank you, t think you mean - only vuln_id_from_tool?

**Review by damiencarol — APPROVED** [review]

**Comment by damiencarol** [comment]
@drJabber  seems good to me but I think you can improve it.
1. more checks in the unit tests (testing more attributes)
2. maybe evaluate switching to use unique_id_from_tool attribute instead of vuln_id_from_tool

**Comment by drJabber** [comment]
@damiencarol , thank you, I feel the code is getting better and better :)
>     1. more checks in the unit tests (testing more attributes)

added extra attributes to unit tests - vuln_id_from_tool, severity, title, description, mitigation. I thing here are enough attributes for test, because there are no more data provided by docker-bench-security

>     2. maybe evaluate switching to use unique_id_from_tool attribute instead of vuln_id_from_tool

added unique_id_from_tool to parser and dedupe config, 
but there are no more unique attributes than "id" in docker-bench-security results, because this id is the section number in CIS Docker benchmark document, so unique_id_from_tool==vuln_id_from_tool=="id"

**Comment by damiencarol** [comment]
this is good add-on. I think the PR is ready

**Review by damiencarol — APPROVED** [review]

**Comment by drJabber** [comment]
Thank you, @damiencarol 
> Just remove the part for vuln_id_from_tool, it could be confusing
Substituted vuln_id_from_tool by unique id from tool

Also fixed unittests

**Review by StefanFl — CHANGES_REQUESTED** [review]
Please add the parser in `docs/content/en/integration/parsers.md`.

**Comment by StefanFl** [comment]
One tiny change request. Great addition, thank you @drJabber

**Comment by drJabber** [comment]
@StefanFl , thank you

> Please add the parser in `docs/content/en/integration/parsers.md`.

Added record in this document

Also fixed unittest

File: dojo/tools/dockerbench/parser.py (hunk 3 above)
Comment (damiencarol) [reply dojo/tools/dockerbench/parser.py]: Yes, you can do that

File: dojo/settings/settings.dist.py (hunk 7 above)
Comment (damiencarol) [reply dojo/settings/settings.dist.py]: yep, but I saw you fixed that

**Comment by damiencarol** [comment]
@StefanFl it's good now, could you review it again?

**Review by damiencarol — APPROVED** [review]

**Comment by drJabber** [comment]
Sorry, modified unittest again - incorrectly manually counted issues.

**Comment by drJabber** [comment]
Hi guys, I invoked unit tests in my local environment, as written in test guide - and there are no issues on it, could you please review it again?

**Review by damiencarol — APPROVED** [review]

**Review by Maffooch — APPROVED** [review]
Wow 4 approvals from @damiencarol must be a solid PR!

@StefanFl leaving the merge to you if all looks good

**Comment by damiencarol** [comment]
> Wow 4 approvals from @damiencarol must be a solid PR!
> 
> @StefanFl leaving the merge to you if all looks good

@drJabber added many changes to improve quality, this explain the multiple reviews :D

**Comment by damiencarol** [comment]
> Wow 4 approvals from @damiencarol must be a solid PR!
> 
> @StefanFl leaving the merge to you if all looks good

@drJabber added many changes to improve quality, this explain the multiple reviews :D

**Review by StefanFl — APPROVED** [review]

**Comment by drJabber** [comment]
Thank you, gentlemen :)

**Pull request merged**